File: 3.longest-substring-without-repeating-characters.java
/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution {
    String substr;
    String longestStr = null;
    public int lengthOfLongestSubstring(String s) {
        for(int i=0; i<s.length(); i++){
            substr = "" + s.charAt(i);
            for(int j=i+1; j<s.length()-i-1; j++){
                if(!substr.contains(s.charAt(j))){
                    substr = substr + "" + s.charAt(j);
                } else {
                    longestStr = longestStr.length > substr.length? longestStr:substr;
                }
        }
    }
    return longestStr.length;
}
}
// @lc code=end

File: 3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_dic = {}
        longest_size = 0
        position = 0

        for i, ch in enumerate(s):
            if ch in char_dic:
                # charが当たるたびに距離を更新
                longest_size = max(longest_size, i-position)
                # 前回のchar時点までpositionを更新する
                position = max(position, char_dic[ch]+1)
            char_dic[ch]=i
        # 最終idx時にはpositionが更新できていない可能性がある
        return max(longest_size, len(s)-position)

if __name__ == "__main__":
    solution = Solution()
    solution.lengthOfLongestSubstring(s="abcabcbb")
        
# @lc code=end

